ratis-server/src/main/java/org/apache/ratis/server/impl/TransferLeadership.java

```diff
@@ -151,29 +150,29 @@ private String tryTransferLeadership(LeaderStateImpl leaderState, Context contex
       return "LogAppender for transferee " + transferee + " is null";
     }
     final FollowerInfo follower = appender.getFollower();
-    final String error = leaderState.sendStartLeaderElection(follower, context.getLeaderLastEntry());
-    if (error == null) {
-      LOG.info("{}: sent StartLeaderElection to transferee {} immediately as it already has up-to-date log",
+    if (follower == null) {
+      return "Follower for LogAppender" + appender + " is null";
+    }
+    if (LeaderStateImpl.isFollowerUpToDate(follower, context.getLeaderLastEntry())) {
+      LOG.info("{}: sending StartLeaderElection to transferee {} immediately as it already has up-to-date log",
+          server.getMemberId(), transferee);
```

Comment:
Let's move the LOG after `tryTransferLeadership(..)` returns.  Then, we can check the return value before log.

Reply:
OK, but tryTransferLeadership returns `null` doesn't mean we have sent StartLeaderElection immediately:

https://github.com/apache/ratis/pull/845/files/b19cccebaf2316fa9d16de21b77b4439f8514d33..d5d6b612c2015ade11e642041c1bdab77728f09a

Reply:
@kaijchen , you are right.  How about adding a `Result` class?  Everything becomes clear and easy to understand.

Reply:
> How about adding a Result class? Everything becomes clear and easy to understand.

OK, let me rebase new changes to https://issues.apache.org/jira/secure/attachment/13056384/845_review2.patch